samples/simulation/qaoa/QAOA.qs

```diff
@@ -0,0 +1,240 @@
+                {
+                    CNOT(target[i], auxiliary);
+                    CNOT(target[j], auxiliary);
+                    Rz(2.0 * time * coupling[6 * i + j], auxiliary);
+                    CNOT(target[i], auxiliary);
+                    CNOT(target[j], auxiliary);
+                }
+            }
+            Reset(auxiliary);
+        }
+    }
+
+    /// # Summary
+    /// Calculate Hamiltonian parameters based on the given costs and penalty
+    ///
+    /// # Input
+    /// ## segmentCosts
+    /// Cost values of each segment
+    /// ## penalty
+    /// Penalty for cases that don't meet constraints
+    ///
+    /// # Output
+    /// ## weights
+    /// Hamiltonian parameters or "weights" as an array where each element corresponds 
+    /// to a parameter h_j for qubit state j.
+    function createHamiltonianWeights(segmentCosts : Double[], penalty: Double) : Double[] {
+        mutable weights = new Double[6];
+        for (i in 0..5) {
+            set weights w/= i <- 4.0 * penalty - 0.5 * segmentCosts[i];
+        }
+        return weights;
+    }
+
+    /// # Summary
+    /// Calculate Hamiltonian coupling parameters based on the given penalty
+    ///
+    /// # Input
+    /// ## penalty
+    /// Penalty for cases that don't meet constraints
+    ///
+    /// # Output
+    /// ## coupling
+    /// Hamiltonian coupling parameters as an array, where each element corresponds
+    /// to a parameter J_ij between qubit states i and j.
+    function createHamiltonianCouplings(penalty: Double) : Double[] {
+        // Calculate Hamiltonian coupling parameters based on the given costs and penalty
+        mutable coupling = new Double[36];
+
+        // Most elements of J_ij equal 2*penalty, so set all elements to this value, 
+        // then overwrite the exceptions
+        for (i in 0..35)
+        {
+            set coupling w/= i <- 2.0 * penalty;
+        }
+        set coupling w/= 2 <- penalty;
+        set coupling w/= 9 <- penalty;
+        set coupling w/= 29 <- penalty;
+
+        return coupling;
+    }
+    
+    /// # Summary
+    /// Perform the QAOA algorithm for this Ising Hamiltonian
+    ///
+    /// # Input
+    /// ## weights
+    /// Instance Hamiltonian parameters or "weights" as an array where each element corresponds to a 
+    /// parameter h_j for qubit state j.
+    /// ## couplings
+    /// Instance Hamiltonian coupling parameters as an array, where each element corresponds
+    /// to a parameter J_ij between qubit states i and j.
+    /// ## timeX
+    /// Time evolution for PauliX operations
+    /// ## timeZ
+    /// Time evolution for PauliX operations
+    operation PerformQAOA(weights : Double[], couplings : Double[], timeX : Double[], timeZ : Double[]) : Bool[] {
+        EqualityFactI(Length(timeX), Length(timeZ), "timeZ and timeX are not the same length");
+
+        // Run the QAOA circuit
+        mutable result = new Bool[6];
+        using (x = Qubit[6])
```

Comment:
There are a lot of magic numbers throughout the sample that depend on there being exactly 6 qubits. Would it be worth making `nQubits` a named variable (perhaps a parameter to this operation), and calculating all of those, such that the number of qubits could easily be changed? Or would that make the code too unreadable?

Reply:
I think I left these in here because there's no native way to create default variables in Q#.
Perhaps it would make sense to make a Python wrapper for this sample?

Reply:
Also, the entire example is supposed to be hard-coded to 6 qubits

Reply:
I've added some docstrings and assertions to explain the magic numbers. Unfortunately can't take it all out unless we want to refactor this sample to generate the QAOA parameters variationally.